Source/Core/VideoCommon/LightingShaderGen.cpp

```diff
@@ -74,19 +74,19 @@ static void GenerateLightShader(ShaderCode& object, const LightingUidData& uid_d
 void GenerateLightingShaderCode(ShaderCode& object, const LightingUidData& uid_data, int components,
                                 u32 numColorChans, const char* inColorName, const char* dest)
 {
-  for (unsigned int j = 0; j < numColorChans; j++)
+  for (u32 j = 0; j < NUM_XF_COLOR_CHANNELS; j++)
   {
     object.Write("{\n");
 
     bool colormatsource = !!(uid_data.matsource & (1 << j));
     if (colormatsource)  // from vertex
     {
-      if (components & (VB_HAS_COL0 << j))
+      if (j < numColorChans && components & (VB_HAS_COL0 << j))
```

Comment:
This can be shortened by reversing the order: <pre>
if (j >= numColorChans)

else if (components & VB_HAS_COL0)

else